x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTaskRandomTests.java

```diff
@@ -0,0 +1,259 @@
+        ShardFollowNodeTask task = createShardFollowTask(concurrency, testRun);
+        startAndAssertAndStopTask(task, testRun);
+    }
+
+    private void startAndAssertAndStopTask(ShardFollowNodeTask task, TestRun testRun) throws Exception {
+        task.start(testRun.startSeqNo - 1, testRun.startSeqNo - 1);
+        assertBusy(() -> {
+            ShardFollowNodeTask.Status status = task.getStatus();
+            assertThat(status.getLeaderGlobalCheckpoint(), equalTo(testRun.finalExpectedGlobalCheckpoint));
+            assertThat(status.getFollowerGlobalCheckpoint(), equalTo(testRun.finalExpectedGlobalCheckpoint));
+            assertThat(status.getIndexMetadataVersion(), equalTo(testRun.finalIndexMetaDataVerion));
+        });
+
+        task.markAsCompleted();
+        assertBusy(() -> {
+            ShardFollowNodeTask.Status status = task.getStatus();
+            assertThat(status.getNumberOfConcurrentReads(), equalTo(0));
+            assertThat(status.getNumberOfConcurrentWrites(), equalTo(0));
+        });
+    }
+
+    private ShardFollowNodeTask createShardFollowTask(int concurrency, TestRun testRun) {
+        AtomicBoolean stopped = new AtomicBoolean(false);
+        ShardFollowTask params = new ShardFollowTask(null, new ShardId("follow_index", "", 0),
+            new ShardId("leader_index", "", 0), testRun.maxOperationCount, concurrency,
+            ShardFollowNodeTask.DEFAULT_MAX_BATCH_SIZE_IN_BYTES, concurrency, 10240,
+            TimeValue.timeValueMillis(10), TimeValue.timeValueMillis(10), Collections.emptyMap());
+
+        BiConsumer<TimeValue, Runnable> scheduler = (delay, task) -> {
+            try {
+                Thread.sleep(delay.millis());
+            } catch (InterruptedException e) {
+                throw new RuntimeException(e);
+            }
+            task.run();
+        };
+        LocalCheckpointTracker tracker = new LocalCheckpointTracker(testRun.startSeqNo - 1, testRun.startSeqNo - 1);
+        return new ShardFollowNodeTask(1L, "type", ShardFollowTask.NAME, "description", null, Collections.emptyMap(), params, scheduler) {
+
+            private volatile long indexMetadataVersion = 0L;
+            private final Map<Long, Integer> fromToSlot = new HashMap<>();
+
+            @Override
+            protected void innerUpdateMapping(LongConsumer handler, Consumer<Exception> errorHandler) {
+                handler.accept(indexMetadataVersion);
+            }
+
+            @Override
+            protected void innerSendBulkShardOperationsRequest(List<Translog.Operation> operations, LongConsumer handler,
+                                                               Consumer<Exception> errorHandler) {
+                for(Translog.Operation op : operations) {
+                    tracker.markSeqNoAsCompleted(op.seqNo());
+                }
+
+                // Emulate network thread and avoid SO:
+                Thread thread = new Thread(() -> handler.accept(tracker.getCheckpoint()));
```

Comment:
I ran these tests and ended with this warning `WARNING: Will linger awaiting termination of 1 leaked thread(s).` I think we need to create a TestThreadPool and tear it down each test.

Reply:
I'm okay with using TestThreadPool instead here, although I think it is not necessary. The threads do eventually stop, otherwise the test framework would have failed.

Reply:
Sorry @martijnvg. I just feel not comfortable when we reply on the test framework to check the termination.